src/gcp/cloudfunctions.ts

```diff
@@ -260,6 +253,73 @@ export async function setIamPolicy(options: IamOptions): Promise<void> {
   }
 }
 
+/**
+ * Formats the service account to be used with IAM API calls, a vaild service account string is
+ * 'service-account@project.iam.gserviceaccount.com', 'service-account@', or 'service-account'.
+ * @param serviceAccount the custom service account created by the user
+ * @param projectId the ID of the current project
+ * @returns a correctly formatted service account string
+ *
+ * @throws {@link Error} if the supplied service account string is empty
+ */
+function formatServiceAccount(serviceAccount: string, projectId: string): string {
+  if (serviceAccount.length === 0) {
+    throw new Error("Service account cannot be an empty string");
+  }
+
+  const suffix = `@${projectId}.iam.gserviceaccount.com`;
+  if (
+    serviceAccount.length > suffix.length &&
+    serviceAccount.slice(serviceAccount.length - suffix.length) === suffix
+  ) {
+    return `serviceAccount:${serviceAccount}`;
+  }
+
+  const emailId =
+    serviceAccount.charAt(serviceAccount.length - 1) === "@"
+      ? `${serviceAccount}${projectId}.iam.gserviceaccount.com`
+      : `${serviceAccount}@${projectId}.iam.gserviceaccount.com`;
+
+  return `serviceAccount:${emailId}`;
+}
+
+/**
+ * Generates the correct IAM Policy for changing the invoker for HTTP triggered functions.
+ * @param projectId the ID of the current project
+ * @param invoker the member(s) that will be granted the invoker role
+ * @returns correctly formatted IAM policy for use in IAM API calls
+ */
+export function generateIamPolicy(
+  projectId: string,
+  invoker: backend.Invoker | backend.Invoker[] = "public"
+): iam.Policy {
+  const invokerRole = "roles/cloudfunctions.invoker";
+  const roleMembers: string[] = [];
+
+  if (typeof invoker === "string") {
+    if (invoker === "public") {
+      roleMembers.push("allUsers");
+    } else if (invoker !== "private") {
+      // if private we do not add any members to the invoker role
+      roleMembers.push(formatServiceAccount(invoker, projectId));
+    }
+  } else {
+    roleMembers.splice(0, 0, ...invoker.map((inv) => formatServiceAccount(inv, projectId)));
+  }
+
+  const policy: iam.Policy = {
+    version: 3,
+    bindings: [
+      {
+        role: invokerRole,
+        members: roleMembers,
+      },
+    ],
+    etag: "",
```

Comment:
It's not great that we're blitzing away any IAM settings by setting etag to "". We should probably merge the `generateIamPolicy` and `setIamPolicy` functions into a `setInvoker` function that would (in a loop that handles race conditions) fetch the existing IAM policy, update only the invoker binding, and then set it with the etag to catch races.